TensorProduct of MatrixSymbols and Trace of them.

johanbluecreek opened the pull request:
The following PR includes the changes that makes it possible to make `TensorProduct` combinations of `MatrixSymbol` objects and take the `trace` of them. The behaviour with this PR is

```
In [1]: import sys

In [2]: sys.path.insert(0, '/local/repo/')

In [3]: from sympy import *

In [4]: from sympy.physics.quantum import TensorProduct as tp

In [5]: n = Symbol('n')

In [6]: X = MatrixSymbol('\\mathbb{X}', n, n)

In [7]: Y = MatrixSymbol('\\mathbb{Y}', n, n)

In [8]: tp(X,Y)
Out[8]: \mathbb{X}x\mathbb{Y}

In [9]: trace(tp(X,Y))
Out[9]: Tr(\mathbb{X})*Tr(\mathbb{Y})

In [10]: tp(X,Y).is_Matrix
Out[10]: True

In [11]: tp(X,Y).is_square
Out[11]: True
```

This PR also includes additions to the corresponding test-files not to break this feature in the future.

Commit:
`TensorProduct` of `MatrixSymbol`

 Added a `args_cnc()` method to `MatrixSymbol` to make it possible to build
`TensorProduct` combinations of them. That is before

```
In [1]: from sympy import *

In [2]: from sympy.physics.quantum import TensorProduct as tp

In [3]: n = Symbol('n')

In [4]: X = MatrixSymbol('\\mathbb{X}', n, n)

In [5]: tp(X,X)

AttributeError: 'MatrixSymbol' object has no attribute 'args_cnc'
```

 While now

```
In [1]: import sys

In [2]: sys.path.insert(0, '/local/repo/')

In [3]: from sympy import *

In [4]: from sympy.physics.quantum import TensorProduct as tp

In [5]: n = Symbol('n')

In [6]: X = MatrixSymbol('\\mathbb{X}', n, n)

In [7]: tp(X,X)
Out[7]: \mathbb{X}x\mathbb{X}
```

Commit:
`Trace` of `TensorProduct`

Added properties `is_Matrix` and `is_square` to `TensorProduct` that are
true when all arguments for `TensorProduct` has `is_Matrix` and `is_square`,
respectively. This gives the desired output for the trace over the
TensorProduct`:

```
In [1]: import sys

In [2]: sys.path.insert(0, '/local/repo/')

In [3]: from sympy import *

In [4]: from sympy.physics.quantum import TensorProduct as tp

In [5]: n = Symbol('n')

In [6]: X = MatrixSymbol('\\mathbb{X}', n, n)

In [7]: Y = MatrixSymbol('\\mathbb{Y}', n, n)

In [8]: trace(tp(X,Y))
Out[8]: Tr(\mathbb{X})*Tr(\mathbb{Y})
```

Furthermore, the following now works on `TensorProduct`:

```
In [9]: tp(X,Y).is_Matrix
Out[9]: True

In [10]: tp(X,Y).is_square
Out[10]: True
```

Commit:
Added the appropriate tests to test_matrix_exprs.py and test_tensorproduct.py
for the new functionality of `MatrixSymbol` and `TensorProduct`.

Commit:
Generalising `is_square` for `TensorProduct`

 As introduced by previous commits, `TensorProduct` has the property
is_square, which previously was only `True` is all arguments of the
`TensorProduct` were square. This is now generalised to if the arguments give
a square matrix, that is if the product of the number of rows of the arguments
equal the product of the columns of the arguments.

 The appropriate tests has been added as well.

Commit:
Merge branch 'master' into matrixsymbols

 Merged a synced master into the 'matrixsymbols'-branch to resolve two
merge-conflicts.

johanbluecreek commented:
There are some issues with the PR as it is now. While one can take `TensorProduct` of `MatrixSymbol` arguments, very limited operations are possible between such `TensorProduct` expressions. I'm working on supplementing this PR with those features as well.

asmeurer commented:
Are more changes forthcoming here?

Ideally we would have a tensor product object outside of the physics module that you could use.

johanbluecreek commented:
@asmeurer,
Yeah, I have some changes that extend the functionality a bit that I will commit when I have cleaned it up a bit. There is however still a problem of Issue #10360 that would still make the it a problem to work with the `TensorProduct` objects made up of `MatrixSymbols`. But I will try and look into that.

I'm not sure I would agree that another tensor product would be good. The physics module one calls the appropriate tensor product for explicit matrices, so it would be natural to have it also support symbolic matrices. There are several areas of physics in which you would like to have the freedom to go from finite sized matrices to infinite matrices/operators as well, so it would be good to have them both as the same type of tensor product object.

To move the tensor product out of the physics module and make it a generic one I would understand, but making several of them would make things messy for certain calculations. But these are just my opinions (and how I would like to use sympy), let me know if I'm out of line.

asmeurer commented:
Moving it would be preferable to duplication. The physics module should only have physics stuff.

Commit:
Added additional functionality

The TensorProduct now also can take `Identity` and `ZeroMatrix` as entries.

Removed the `test_args_cnc` since it is not a completely clean fix and is
instead implicit from the test for `TensorProduct`.

Also added tests for the `Identity` and `ZeroMatrix` as arguemts for the
`TensorProduct`.

johanbluecreek commented:
From the latest commit `Identity` and `ZeroMatrix` now is supported as entries to the `TensorProduct` as well, and it also handles `MatMul` expressions correctly. Tests were also added.

To further develop the functionality of `TensorProduct` with `MatrixSymbol` arguments, Issue #10360 needs to be resolved. But this is beyond the scope of this PR, and this PR is finished.

asmeurer commented on sympy/matrices/expressions/matmul.py:
```diff
@@ -10,9 +10,10 @@
 
 
-class MatMul(MatrixExpr):
+class MatMul(MatrixExpr, AssocOp):
```
What required this change?

asmeurer replied:
I ask because AssocOp subclasses must define `Identity`. Sadly, the way the class is defined, it's a class attribute (and even if it were an instance attribute, a MatMul has two identities, left and right). I think unless we make some modifications, it isn't possible to use AssocOp with MatMul.

johanbluecreek replied:
Regarding `AssocOp`, I added that because the error:

"AttributeError: type object 'MatMul' has no attribute '_from_args'"

was triggered. `Mul` gets its `_from_args` from `AssocOp` hence I added it here. I can look into a another fix for this if this is overkill.

Regarding the "MatMul has two identities", please inform me about what I'm missing, but a `MatrixSymbol` only has the `Identity`, which is (as I understand it) a symbolic representation of the identity matrix (with the integer `1` on the diagonal), which has the property (for any matrix A (This case A is a MatrixSymbol and I is the Identity of )):
 `I*A = A*I =A`
hence the same action from left and right.

asmeurer replied:
Consider for instance `MatMul(MatrixSymbol("A", n, m), MatrixSymbol("B", m, k))`. 

Even ignoring that, one of the reasons AssocOp wants an identity is so that things like `Mul()` with no arguments could return the identity. But there's no way for `MatMul()` to return an identity (it needs a shape).

johanbluecreek replied:
Ah, of course. I will look into another fix then.

asmeurer commented on sympy/physics/quantum/tensorproduct.py:
```diff
@@ -116,6 +117,17 @@ class TensorProduct(Expr):
 
+    @property
+    def is_Matrix(self):
+        return all(arg.is_Matrix for arg in self.args)
```
I think this was discussed in some other issue, but it would be more useful to have a KroneckerProduct MatrixExpression class. That way you could have first-class support for operations with other matrices, and things like evaluation on explicit matrices.

johanbluecreek replied:
Valid point. But I'm not sure how this implementation would work. For example, as a physicist, one would desire to use a `MatrixSymbol` of size `NxN` in which `N \to \infty`, which would mean that they are truly operators. But to have the freedom to interpolate between operators and "large `N`" matrices would really be desired for some branches of physics.

Separating "large `N`" matrices and operators to `KroneckerProduct` and `TensorProduct`, respectively,  would make this interpolation difficult (or?). This is my only worry. (is it the Issue #10771 you are thinking of?)

Then there is also the issue with `paulialgebra`, these are currently implemented as non-commutative `Symbol` objects, but are really matrices (most commonly represented by `2x2`matrices, and with the alternative implementation to PR #11598 (mentioned in the same PR) one would need to convert them to be `MatrixSymbol` objects instead.). These are often considered in tensor product combinations when it comes to, for example, Killing spinor equations.

My vision was to convert the current `paulialgebra` to `MatrixSymbol` objects so that these could be of more general use.

johanbluecreek commented:
If, @moorepants, the `PR: sympy's turn` tag has been added since this is moving towards a merge, I just want to note that I'm working on resolving the issue regarding the AssocOp mentioned by @asmeurer, to avoid the reliance on it.

moorepants commented:
I've switched it to "author's turn".

Commit:
Removed the use of AssocOp

 Instead of using AssocOp for MatMul, the _from_args() is taken from AssocOp
and used instead. This avoids the problems of the assumed existence of an
identity, as mentioned by @asmeurer.

Commit:
Merge branch 'master' into matrixsymbols

 Synced to upsteam and merging

johanbluecreek commented:
@moorepants, @asmeurer, this can now be set to "sympy's turn".

gxyd commented:
@siefkenj would you like to review this PR? This is related to matrices module.

siefkenj commented on sympy/matrices/expressions/matexpr.py:
```diff
@@ -422,6 +422,12 @@ def __call__(self, *args):
         return MatrixElement(self, i, j)
 
+    def args_cnc(self):
+        return [[],[self]]
```
Do these need to be lists instead of tuples?

siefkenj commented on sympy/core/trace.py:
```diff
@@ -167,7 +167,13 @@ def doit(self, **kwargs):
         """
         if hasattr(self.args[0], '_eval_trace'):
-            return self.args[0]._eval_trace(indices=self.args[1])
+            if self.args[1] == ():
```
When is `args[1]` an empty list? What error does `_eval_trace` give if `indices==()`?

I think this would be cleaner as

```
if self.args[1]:
    return self.args[0]._eval_trace(indices=self.args[1])
try:
    return self.args[0]._eval_trace()
except NotImplementedError:
    return Tr(self.args[0].doit())
```

siefkenj reviewed (commented):
I think this PR needs to be rebased with master.

mijo2 commented:
I will come up with a PR for this soon.
